src/sources/docker.rs

```diff
@@ -445,17 +451,46 @@ impl EventStreamBuilder {
         main_send: UnboundedSender<ContainerLogInfo>,
     ) -> Self {
         EventStreamBuilder {
-            core,
+            core: Arc::new(core),
             out,
             main_send,
         }
     }
 
     /// Constructs and runs event stream
-    fn start(&self, id: ContainerId) -> ContainerState {
-        let mut state = ContainerState::new(id, self.core.now_timestamp);
-        self.restart(&mut state);
-        state
+    fn start<O: Into<Option<ContainerMetadata>>>(
+        &self,
+        id: ContainerId,
+        metadata: O,
+    ) -> ContainerState {
+        if let Some(metadata) = metadata.into() {
+            self.start_event_stream(ContainerLogInfo::new(id, metadata, self.core.now_timestamp));
+        } else {
+            let metadata_fetch = self
+                .core
+                .docker
+                .containers()
+                .get(id.as_str())
+                .inspect()
+                .map_err(|error| error!(message="Fetching container details failed",%error))
+                .and_then(|details| {
+                    ContainerMetadata::from_details(&details)
+                        .map_err(|error| error!(message="Metadata extraction failed",%error))
+                });
+
+            let this = self.clone();
+            let create = metadata_fetch.map(move |metadata| {
+                this.start_event_stream(ContainerLogInfo::new(
```

Comment:
Can we make it so `start_event_stream` runs on the same task as the metadata fetch? This would just need to be changed to a `and_then` and make `start_event_stream` return a future.